### PR TITLE
Machine-readable reports

### PR DESCRIPTION
**Pull request by samoylovfp**
Solves https://github.com/tag1consulting/goose/issues/499 (once finished)

The idea is to have reports written into a json file,
so further processing and merging can be made out of the process

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -124,6 +125,44 @@ pub fn get_worker_id() -> usize {
...
+        let cmd = self.command_line_option.clone();
+
+        file.write_all(report)
+            .await
+            .map_err(|e| GooseError::InvalidOption {
+                option: cmd,
+                value: path,
+                detail: format!("Failed to write to report file: {}", e),
+            })?;
+        // FIXME: should we add context here as well?
```
Comment (samoylovfp) [inline src/lib.rs:160]: @jeremyandrews I would appreciate feedback on this

File: src/lib.rs [hunk 2]
```diff
@@ -124,6 +125,44 @@ pub fn get_worker_id() -> usize {
...
+                option: self.command_line_option.clone(),
+                value: self.path.clone(),
+                detail: format!("Failed to create report file: {}", e),
+            })?;
+        self.file = Some(file);
+        Ok(())
+    }
+
+    pub(crate) async fn write_all_and_flush(&mut self, report: &[u8]) -> Result<(), GooseError> {
+        let file = self.file.as_mut().unwrap();
```
Comment (jeremyandrews) [inline src/lib.rs:149]: I'm leery of having an `.unwrap()` here. At minimum, please add a comment explaining why this is safe.

File: src/lib.rs (hunk 1 above)
Comment (jeremyandrews) [reply src/lib.rs:160]: It's definitely helpful to add context anywhere we can fail. I imagine flush can fail for example if a filesystem fills up, and that's definitely something that can happen during a large load test.

File: src/lib.rs [hunk 3]
```diff
@@ -124,6 +125,44 @@ pub fn get_worker_id() -> usize {
 /// Socket used for coordinating a Gaggle distributed load test.
 pub(crate) struct Socket {}
 
+/// Wrapper over File object that contains context for errors
```
Comment (jeremyandrews) [inline src/lib.rs:128]: Perhaps instead of moving this all into `lib.rs`, it would be more appropriate for it to live in `report.rs`?

**Review by jeremyandrews — CHANGES_REQUESTED** [review]
So far it's looking good. Adding more context to errors is always helpful. A few little comments, I'm looking forward to seeing the machine-readable aspects added.

**Comment by samoylovfp** [comment]
Didn't have time to get to it this week and I am on holiday next 3 weeks. I expect I'll return to this some time in the next 40 days

File: src/lib.rs (hunk 1 above)
Comment (samoylovfp) [reply src/lib.rs]: Added

File: src/lib.rs (hunk 2 above)
Comment (samoylovfp) [reply src/lib.rs]: Removed the filed, the files is now created here

File: src/lib.rs (hunk 3 above)
Comment (samoylovfp) [reply src/lib.rs]: Good idea. Moved there

**Comment by samoylovfp** [comment]
We found out we can extract the need information from requests log.

**Comment by mTsBucy1** [comment]
@samoylovfp
May I ask how you further processed the data? Did you parse the json requests log via code? Or did you use a Spreadsheet application?

Because I am currently in the process of using the csv requests log (which lead to #515)  and importing in Excel, where I can do more customized processing, and I was wondering what other paths there might be.

**Comment by samoylovfp** [comment]
We wrote our own small app using egui, that loads different benchmarking results and shows request time histograms on a log scale. Our app has several caching layers and performance of these layers show up as several "humps" on the histograms, this helped us track changes to the overall performance as well as each cache layer

**Pull request closed without merging**